### ISSUE TITLE
file_rec input line bug

### ISSUE THREAD

**Issue by FinallyFinancialFreedom** [Shougo/denite.nvim#184]
**Warning:  I will close the bug issue without the minimal init.vim and the reproduce ways.**

# Problems summary
my configuration:
```vim
call denite#custom#source('file_mru', 'matchers', ['matcher_fuzzy', 'matcher_ignore_globs'])
```
when i start up vim8, trigger it at home dir, ok i know with neomru.vim, but it's wierd
for example, when i press `di` to search file with name 'di', it will show files match 'di', but the input line only show 'd', then i press backspace, it will show files match 'd',but the input line show di.
this error is in file_rec and file_mru of MacVim with vim8.0, but neovim in terminal is working fine.
## Environment Information (Required!)

 * denite version(SHA1):newest

 * OS:MacOS

 * Vim/neovim version:MacVim with vim8.0, and neovim in terminal is ok.

 * `:CheckHealth` result(neovim only):

## Provide a minimal init.vim with less than 50 lines (Required!)

```vim
" Your minimal init.vim
set runtimepath+=~/path/to/denite.nvim/
```

## The reproduce ways from neovim starting (Required!)

 1. foo
 2. bar
 3. baz

**Comment by Shougo** [comment]
The lag is not bug.
It is wired currently though.

**Comment by Shougo** [comment]
And Vim8 is slow than neovim.

**Comment by lambdalisue** [comment]
Ping me after 3/8 while I'm a bit busy for other things.

**Comment by jwcxz** [comment]
I have this same issue. It's specific to MacVim. The issue is that the prompt display lags behind by exactly one key press. 

If you type "di", you will only see "d". If you hit backspace, you will now see "di". If you type "a", you will now see "d". If you type "e", you will now see "da", etc.

**Comment by jwcxz** [comment]
(Vim8 in terminal works fine for me -- this issue does not occur)

**Comment by Shougo** [comment]
> Ping me after 3/8 while I'm a bit busy for other things.

OK.  Don't mind.

> I have this same issue. It's specific to MacVim. The issue is that the prompt display lags behind by exactly one key press.

I cannot reproduce it in Linux GVim.
It seems MacVim only issue.

**Comment by jwcxz** [comment]
I made an interesting discovery here.  Not anywhere close to root-cause, but it's something.

It seems to at least partially have to do with MacVim not redrawing the prompt after keypresses (whereas  in terminal, a redraw occurs immediately).

For example, try changing `rplugin/python3/denite/ui/prompt.py:on_keypress()` to do this:

```python
    def on_keypress(self, keystroke):
        m = ACTION_KEYSTROKE_PATTERN.match(str(keystroke))
        if m:
            r = self.action.call(self, m.group('action'))
            self.redraw_prompt()
            return r
        elif self.denite.current_mode == 'insert':
            # Updating text from a keystroke is a feature of 'insert' mode
            self.update_text(str(keystroke))
            self.redraw_prompt()
```

Observe that I added two calls to `self.redraw_prompt()`.  Now, the prompt will update quickly after a keystroke, not be one character behind.

I'm not familiar with the code enough yet to understand why there's an instantaneous redraw when using terminal vim (or gvim) that doesn't exist when using MacVim.

Note: this *still* doesn't solve the issue that the prompt feels very laggy on MacVim.  Typing quickly results in slow redraws.  Using the arrow keys to scroll through the list of matches is also very slow to redraw.  Surprisingly, in the terminal, it's much faster.  I can't seem to get any useful information with vim's `profile` tool as to what's taking so long unfortunately.  :(

**Comment by Shougo** [comment]
Fixed by #231.
